lib/logstash/filters/grok/timeout_enforcer.rb

```diff
@@ -72,31 +68,16 @@ def start_thread_groking(thread)
     @threads_to_start_time.put(thread, java.lang.System.nanoTime)
   end
 
-  # Returns falsy in case there was no Grok execution in progress for the thread
-  def stop_thread_groking(thread)
-    @threads_to_start_time.remove(thread)
-  end
-
   def cancel_timed_out!
     now = java.lang.System.nanoTime # save ourselves some nanotime calls
-    @threads_to_start_time.entry_set.each do |entry|
-      start_time = entry.get_value
-      if start_time < now && now - start_time > @timeout_nanos
-        thread  = entry.get_key
-        # Ensure that we never attempt to cancel this thread unless a Grok execution is in progress
-        # Theoretically there is a race condition here in case the entry's grok action changed
-        # between evaluating the above condition on the start_time and calling stop_thread_groking
-        # Practically this is impossible, since it would require a whole loop of writing to an
-        # output, pulling new input events and starting a new Grok execution in worker thread
-        # in between the above `if start_time < now && now - start_time > @timeout_nanos` and
-        # the call to `stop_thread_groking`.
-        if stop_thread_groking(thread)
-          @cancel_mutex.lock
-          begin
-            thread.interrupt()
-          ensure
-            @cancel_mutex.unlock
-          end
+    @threads_to_start_time.forEach do |thread, start_time|
+      # Use compute to lock this value
+      @threads_to_start_time.compute(thread) do |thread, start_time|
+        if start_time && start_time < now && now - start_time > @timeout_nanos
+          thread.interrupt
+          nil # Delete the key
```

Comment:
a little unsure about the jruby/java boundary here -- `nil` is not `null`, and while I can find documentation that the coersion happens from [Java to Ruby](https://github.com/jruby/jruby/wiki/CallingJavaFromJRuby#java-to-ruby), I'm not finding anything explicit about the other direction.

Reply:
```
$ irb
irb(main):001:0> require "java"
=> false
irb(main):002:0> h = java.util.concurrent.ConcurrentHashMap.new
=> {}
irb(main):003:0> h.inspect
=> "{}"
irb(main):004:0> h.put("a", 1)
=> nil
irb(main):005:0> h.inspect
=> "{\"a\"=>1}"
irb(main):006:0> h.compute("a") {|v| nil}
=> nil
irb(main):007:0> h.inspect
=> "{}"
```

Reply:
Thanks for confirming this behavior colin!